View Keybindings Widget

vince-fugnitto opened the pull request:
**Keybindings Widget** used to `view`, `edit` and `search` keybindings

----------

**View**

![selection_013](https://user-images.githubusercontent.com/40359487/45622130-ee1f9f80-ba50-11e8-9891-8eac498b6ccf.png)

**Edit**
![selection_029](https://user-images.githubusercontent.com/40359487/46036971-de4d2e80-c0d4-11e8-97e6-0115fbdae482.png)

**Fuzzy Search With Highlighting**

![selection_024](https://user-images.githubusercontent.com/40359487/45897661-c3617e00-bda5-11e8-80d5-001bc48805de.png)

![selection_016](https://user-images.githubusercontent.com/40359487/45642732-1de99a00-ba87-11e8-85fc-3715f13d1913.png)

Signed-off-by: Vincent Fugnitto <vincent.fugnitto@ericsson.com>

<!-- Please provide a clear and meaningful description to the CHANGELOG.md file if this PR contributes some significant changes -->

vince-fugnitto commented on packages/keymaps/src/browser/keymaps-frontend-contribution.ts:
```diff
@@ -24,31 +24,60 @@ import {
         id: 'keymaps:open',
         label: 'Open Keyboard Shortcuts'
     };
+    export const OPEN_VIEW_KEYBINDINGS: Command = {
```
I agree but for the initial PR (**view** and **search**), we won't have a way yet as to edit the keybindings if we re-implement the `OPEN_KEYMAPS` command. Until the second PR of editing keybindings through this UI is complete I'd keep the two, but its up to you

akosyakov replied:
It would be nice to merge in the master something what is complete, if it is not complete it should not be accessible for users. // cc @svenefftinge

vince-fugnitto replied:
I thought it was more just a question of the scope of the PR. The first PR would include simply **viewing** and **searching** all available keybindings (`View Keybindings`). Subsequent PRs would offer more functionality like the ability to **edit keybindings**, **reset keybindings to default**, **sort** and will most likely remove the `OPEN_KEYMAPS` as seen today.

However, if you believe that it should simply be one PR which includes `viewing`, `editing`, `searching`, `sorting`, along with other minor enhancements than I don't object to moving this PR to WIP.

vince-fugnitto replied:
Thanks, my initial goal was in fact to have the editing functionality as part of my initial PR so I'll work on getting it complete along with addressing all the comments you provided :)

vince-fugnitto commented on packages/keymaps/src/browser/keybindings-widget.tsx:
```diff
@@ -0,0 +1,202 @@
+        </React.Fragment>;
+    }
+
+    protected renderKeybinding(keybinding?: string): React.ReactNode {
```
I am using monaco styling, just that I felt like vscode, black keybindings looked better on the dark theme to me. Perhaps I can change the global style so that all our keybindings across the platform are similiar

akosyakov replied:
I mean, for instance, Monaco displays some keys like locale-specific icons, i.e. ctrlcmd, alt, shift and so on.

vince-fugnitto replied:
I've been thinking of doing that as well, but perhaps in an additional PR and it can be re-used across Theia

akosyakov replied:
Sounds good, maybe you can open an issue with a list of all features which should be added later?

vince-fugnitto replied:
Sure will do!

vince-fugnitto commented:
@akosyakov thanks for the feedback! Still learning React! :) I'll make the updates

vince-fugnitto commented on packages/keymaps/src/browser/keybindings-widget.tsx:
```diff
@@ -0,0 +1,202 @@
+        return document.getElementById('search-keybindings') as HTMLInputElement;
+    }
+
+    protected searchFilter(item: KeybindingItem, query: string): boolean {
```
@akosyakov I'm using the fuzzy package as of my latest changes but was having a lot of difficulty processing the data afterwards (regexp, building html from my data) in order to adjust the formatting of the table to display results in `bold` and `blue`. I'm wondering if there is a simple way to do it without any performance degradation.

![selection_023](https://user-images.githubusercontent.com/40359487/45840382-d2ccc280-bce4-11e8-9370-da416e25126c.png)

akosyakov replied:
Let's leave it out for a separate PR? If you want to me look how it can be optimized please push your changes.

vince-fugnitto replied:
I can include fuzzy searching but perhaps keep the highlighting as part of another PR if it can't be optimised. I pushed the changes, but please note the code might be really ugly as of now :)

akosyakov commented on packages/keymaps/src/browser/keybindings-widget.tsx:
```diff
@@ -0,0 +1,268 @@
+ ********************************************************************************/
+
+import React = require('react');
+import * as fuzzy from 'fuzzy';
```
`"fuzzy": "^0.1.3"` please add it to package.json

vince-fugnitto commented on packages/keymaps/src/browser/keybindings-widget.tsx:
```diff
@@ -0,0 +1,268 @@
+        const items = this.getItems();
+        items.forEach(item => {
+            const keys: (keyof KeybindingItem)[] = ['id', 'command', 'context', 'keybinding'];
+            let matched = false;
```
my only concern is that if I early exit for instance because a value in `command` matched, other columns for that same row won't get fuzzy highlighting.

akosyakov replied:
ok, makes sense

vince-fugnitto commented:
@akosyakov thanks for noticing the bug! I updated the regexp and processing so that it works correctly now. I also got the fuzzy highlighting working on all columns now especially `keybindings`.

akosyakov commented:
@vince-fugnitto Do you mean is it ready for the review? It is still marked as WIP.

vince-fugnitto commented:
@akosyakov I've added editing `keybindings` to the PR. Initially I will only support editing the actual keybinding and not the `context` as well since I will need to create my own dialogue which can support multiple input fields. I've also added validation to check if its a valid keybinding and does not collide with any existing ones.

akosyakov commented on packages/keymaps/src/browser/keybindings-widget.tsx:
```diff
@@ -0,0 +1,321 @@
+            <div className='search-kb-container'>
+                <input id='search-kb'
+                    className={(this.items.length > 0) ? '' : 'no-kb'}
+                    type='text' placeholder='Search keybindings' onKeyUp={debounce(() => this.searchKeybindings(), 50)}></input >
```
it creates a new debounce function each time and does not take any effect, you should store a function in the property, i.e
```ts
protected readonly searchKeybindings = debounce(() => this.doSearchKeybindings(), 50);
protected doSearchKeybindings(): void {
    ...
}
```

vince-fugnitto commented:
@akosyakov I noticed that its an existing issue with keybindings, where if we add a keybinding to `keymaps.json`, it successfully adds but does not replace an existing binding, therefore the old one will work as well. Perhaps it should be fixed in a later issue independent of the PR?

vince-fugnitto commented:
@akosyakov I've also added the `revert` as you mentioned since it was relatively simple to implement for this PR. I would however like to check first if the keybinding exists in the json before creating the revert button, but went for the message notification (if no custom keybinding is present) instead to not change too much to accommodate for the async call.

akosyakov commented on packages/keymaps/src/browser/keymaps-service.ts:
```diff
@@ -53,20 +63,40 @@ export class KeymapsService {
+            const item: KeybindingJson = { 'command': command, 'keybinding': keybinding };
+            keybindings.push(item);
+        }
+        await this.resource.saveContents(JSON.stringify(keybindings, undefined, 4).concat('\n'));
```
I see, that's unfortunate :( please add an issue to list that we should try to preserve the original content.

vince-fugnitto commented:
@marechal-p thanks! should be fine now
